add School dropdown to workshop signup flow

davidsbailey opened the pull request:
This implements a whole bunch of UI functionality to the workshop sign-up / join flows as listed in this spec, most notably by adding the school dropdown: https://docs.google.com/document/d/1rIp1hbkuYUlg4lJqf5NhBnpIPH9Qsozga2dmefaarZw/edit#heading=h.snfw7lsx5jz2 this is NOT supposed to be pixel-perfect, however it is supposed to add all of the functionality shown in the workshop section of the spec.

screencast:
![school-dropdown-adhoc](https://cloud.githubusercontent.com/assets/8001765/20114441/33b6845a-a5a9-11e6-9c94-1cc9f76fe7c7.gif)

collateral damage:
* changed the school info model verification to accommodate some changes to the spec around when state and zip should be present
* if you signed up for a workshop before this change goes in, and you join the section after, you will have to re-enter your school information. Given the size of this PR I'd rather merge it as is rather than add a whole bunch of logic to try and move the old data format to the new one, plus we have a captive audience at the workshops so we can be sure this step won't discourage anyone from joining the section. If we want a better user experience we can add it in a follow-up PR.

manually verified:
* enroll using old data format, join using new data format
* enroll using new data format, join using new data format
* join using new data format with existing no school info (this is not uncommon in the production db, presumably from an earlier version of this feature which did not require school info)

to verify:
- [x] dropdown performance on chef-managed machines (very slow on localhost)

davidsbailey commented:
Andrew, apologies for the week long code review. Can you please review? Please let me know if I can do anything to make reviewing easier. 

@breville I would like your input also. 

FYI @ashercodeorg @ewjordan

davidsbailey commented:
Here is a way you can see clearer diffs on the files which were moved in the first commit of this PR: https://github.com/code-dot-org/code-dot-org/compare/school-dropdown-base...school-dropdown?expand=1

davidsbailey commented:
You can see it working here: https://dashboard-adhoc-school-dropdown.cdn-code.org/pd/workshops/1/enroll

breville commented on dashboard/app/views/shared/_school_info.html.haml:
```diff
@@ -0,0 +1,107 @@
+        %input#school-district-id.form-control{name: "school_district_id_source", type: "text", required: false}
+
+        -# This is the input we'll actually submit as part of the form.  We'll copy the Selectized value into
+        -# here whenever it's changed.  Sometimes we'll override this with code, say 0 for "unknown district".
```
I'm not sure whether this comment about 0 is still applicable or not.

breville commented:
This is an epic change, and from a read-through I didn't see any obvious issues.  In my experience it all comes down to testing every UI path and having lots of tests, both of which you seem to have done well.

One observation from playing with the UI is that we have the control at the bottom of the page and we keep expanding it below the bottom, making the user play catch up (scrolling down) with the options as they expose them.  I wonder if there's a way to pull the control up much higher in the page, perhaps through putting it in a second column?

poorvasingal commented:
Dave - this is looking great :D  I emailed you with some feedback on a couple of small things. 

@breville - That's a good point on needing to scroll.  I don't think we should worry about changing the layout right now because we are thinking of doing an update here anyway in the future to help with that, where we put the field names and the input fields side by side (instead of stacked).  That way, we can have consistent UI across the board for the feature (in sign up, accounts page, and the workshop registration page).

davidsbailey commented:
Yeah the scrolling is super annoying. Because the new mocks show a much more compact layout, I'm hoping the problem will go away when we move to it.

davidsbailey commented:
Hi @poorvasingal , I've addressed some of your feedback:
1. school and district dropdowns are now both ordered by school name. see screencast:
![school-dropdown-alphabetize](https://cloud.githubusercontent.com/assets/8001765/20124114/efe3b3e4-a5d8-11e6-9a09-67b5ddf4f393.gif)
2. the highlighting issue is a weird one. it's actually a bug in the `selectize` tool we are using. you can see the same bug here: https://selectize.github.io/selectize.js/ --> "Single Item Select". There are workarounds, but I think they will be too risky to add to this PR, especially since this is not actually a regression (the school DISTRICT dropdown already behaved like this).

technical notes:
to fix `selectize` highlighting, I'd recommend waiting for the fix (https://github.com/selectize/selectize.js/pull/1169) to go in and then update our `rails-selectize` to point to the new version. FWIW it looks like these two dropdowns are the only part of our system using `rails-selectize`. @brendan please let me know if you disagree.

davidsbailey commented:
@breville @aoby I've removed the unneeded form fields associated with the school and district dropdowns. Please let me know if you have any further comments.

poorvasingal commented:
Thanks @davidsbailey -  I agree with your recommendation.

aoby commented on apps/src/sites/studio/pages/schoolInfo.js:
```diff
@@ -6,6 +6,17 @@ window.SchoolInfoManager = function (existingOptions) {
+    }
+    return 0;
+  }
+
```
Why this method? Can't you use selectize's [sortField](https://github.com/selectize/selectize.js/blob/master/docs/usage.md) property? I know very little about selectize so maybe that won't work here?

davidsbailey replied:
Nice catch. Your reading is correct, `sortField` works great. I didn't see it because I was looking at the [api docs](https://github.com/selectize/selectize.js/blob/master/docs/api.md) which misleadingly doesn't document much of the api.

aoby commented on dashboard/app/views/shared/_school_info.html.haml:
```diff
@@ -64,13 +58,7 @@
-        -# Selectize was showing this 0 as a character when it was disabled, hence giving Selectize its own
-        -# input to work with above.
-        %input#school-id-form.form-control{name: "#{form_name}[school_id]", type: "hidden", required: false}
+        %input#school-id.form-control{name: "#{form_name}[school_id]", type: "text", required: false}
```
Much cleaner! Nice

davidsbailey commented:
@aoby I've fixed the `sortField` implementation. @poorvasingal after getting irritated at the scroll behavior while trying to manually test the feature, I've made it so that we scroll to the bottom whenever the dropdown is loading or opened. Here's how it looks:
![school-dropdown-scrolling](https://cloud.githubusercontent.com/assets/8001765/20159433/6bd6aa62-a695-11e6-99f0-877631277b47.gif)
